core/finality-grandpa/src/communication/gossip.rs

```diff
@@ -1090,45 +1175,127 @@ impl<Block: BlockT> network_gossip::Validator<Block> for GossipValidator<Block>
 			} else {
 				false
 			};
-
-			// downgrade to read-lock.
-			(RwLockWriteGuard::downgrade(inner), do_rebroadcast)
+			(inner, do_rebroadcast)
 		};
 
 		Box::new(move |who, intent, topic, mut data| {
 			if let MessageIntent::PeriodicRebroadcast = intent {
 				return do_rebroadcast;
 			}
 
-			let peer = match inner.peers.peer(who) {
-				None => return false,
-				Some(x) => x,
-			};
-
-			// if the topic is not something we're keeping at the moment,
+			// If the topic is not something we're keeping at the moment,
 			// do not send.
-			let (maybe_round, set_id) = match inner.live_topics.topic_info(&topic) {
-				None => return false,
-				Some(x) => x,
+			let (maybe_round, set_id) = {
+				match inner.peers.peer(who) {
+					None => return false,
+					Some(_) => {},
+				};
+				match inner.live_topics.topic_info(&topic) {
+					None => return false,
+					Some(x) => x,
+				}
 			};
 
-			// if the topic is not something the peer accepts, discard.
+			let message = GossipMessage::<Block>::decode(&mut data);
+
 			if let Some(round) = maybe_round {
-				return peer.view.consider_vote(round, set_id) == Consider::Accept
-			}
+				if let Some(GossipMessage::VoteOrPrecommit(msg)) = message {
+					// If we don't know about the peer,
+					// or if the topic is not something the peer accepts, discard.
+					match inner.peers.peer(who) {
+						None => return false,
+						Some(peer) => {
+							if !(peer.view.consider_vote(round, set_id) == Consider::Accept) {
+								return false
+							}
+						},
+					}
 
-			// global message.
-			let local_view = match inner.local_view {
-				Some(ref v) => v,
-				None => return false, // cannot evaluate until we have a local view.
-			};
+					let (primary_proposals_to_peer, prevotes_to_peer, precommits_to_peer) = {
```

Comment:
The `message_allowed` method may be called more than once with the same message, so we can't update the tallies the way we're doing here. All of the messages we see here we should have already seen through `validate` previously, except for the messages we send ourselves (which we will see here but not in `validate`).

Ideally we wouldn't need to do anything with the tallies here, but we have to update the tallies with the messages we send ourselves.

Reply:
Yes, in `message_allowed` we only tally what we sent out, and in `validate` we only tally what we receive. 

> The message_allowed method may be called more than once with the same message

I think the issue here might be that it's used as part of `validate` to "broadcast topics", and then it's called as part of `propagate` in various places. 

Do you mean that in some cases we shouldn't tally what we send out, for example when broadcasting as part of `validate`. 

How can `message_allowed` "known" when an outgoing message should be tallied(when it comes from the node itself?), and when it shouldn't be tallied?